src/CBLDatabase_Internal.hh

```diff
@@ -351,48 +352,40 @@ private:
 
     void callDocListeners();
     
-    void stopActiveStoppables() {
-        std::unordered_set<CBLStoppable*> stoppables;
+    void stopActiveService() {
+        std::unordered_set<CBLActiveService> services;
         {
             std::lock_guard<std::mutex> lock(_stopMutex);
             if (_stopping)
                 return;
             _stopping = true;
-            stoppables = _stoppables;
+            services = _services;
         }
-        
+
         // Call stop outside lock to prevent deadlock:
-        for (auto s : stoppables) {
-            s->stop();
+        for (const auto& s : services) {
+            s.stop();
         }
         
         std::unique_lock<std::mutex> lock(_stopMutex);
-        if (!_stoppables.empty()) {
+        if (!_services.empty()) {
             CBL_Log(kCBLLogDomainDatabase, kCBLLogInfo,
                     "Waiting for %zu active replicators and live queries to stop ...",
-                    _stoppables.size());
-            _stopCond.wait(lock, [this] {return _stoppables.empty();});
+                    _services.size());
+            _stopCond.wait(lock, [this] { return _services.empty(); });
```

Comment:
I am wondering what would happen if the listener is stopped from from different thread from the Listener itself. Could the listener be deleted when you try to unregister from here.

Reply:
>  if the listener is stopped from from different thread from the Listener itself.
I would expect that the listener will still be retained as users shouldn't release it then stop?

Reply:
Maybe I don't understand your question.

Reply:
The database retained the listener and the retain cycle is broken when the listener is stopped. So there shouldn't be the issue. Also as long as the listener's stop() is thread-safe, users can stop on a different thread at the same time that the database is stopped. The unregisterService() is thread safe and the listener is retained so there shouldn't be the issue.